tests/models/encoder_decoders/wikitables_semantic_parser_test.py

```diff
@@ -25,6 +24,82 @@ def setUp(self):
     def test_model_can_train_save_and_load(self):
         self.ensure_model_can_train_save_and_load(self.param_file)
 
+    def test_correct_neighbor_indexes(self):
+        # Verify that 2nd entity has correct neighbor indexes.
+        worlds = []
+        for instance in self.dataset.instances:
+            worlds.append(instance.fields['world'].metadata)
+        num_entities = len(worlds[0].table_graph.entities)
+        tensor = Variable(torch.LongTensor([]))
+        neighbor_indexes = self.model._get_neighbor_indexes(worlds, num_entities, tensor)
+        assert neighbor_indexes.size(1) == num_entities
+        assert neighbor_indexes.size(2) == num_entities
+        lst = [0]*(num_entities)
+        lst[0], lst[1] = 45, 45
+        second_entity_neighbors = (torch.LongTensor(lst))
+        assert torch.equal(neighbor_indexes[0][1].data, second_entity_neighbors)
+
+    def test_all_types_present_in_type_vector(self):
+        # Verifies that at least 1 instance of every one hot vector
+        # is present. This guards against a bug where entity equality
+        # wasn't working since each entity is of type EntityType and
+        # not str. The result of this bug was the same one hot vector
+        # always picked regardless of type.
+        worlds = []
+        for instance in self.dataset.instances:
+            worlds.append(instance.fields['world'].metadata)
+        num_entities = max([len(world.table_graph.entities) for world in worlds])
+        tensor = Variable(torch.LongTensor([]))
+        type_vector, _ = self.model._get_type_vector(worlds, num_entities, tensor)
+        # (batch_size, num_types)
+        sums = torch.sum(type_vector, dim=1)
+        x = torch.nonzero(sums.data)
+        # this means every element in sums is nonzero hence both 2 numbers
+        # representing its index will be in x
+        assert x.size(0) * x.size(1) == 2 * sums.size(0) * sums.size(1)
+
+    def test_get_prob_entity_valid_probability_distribution(self):
```

Comment:
Also here, this would be better to just pick some numbers in a small `linking_scores` tensor, make the necessary masks by hand, and make sure that the values come out normalized the way you expect.  I'd be a lot more confident that the method is doing the right thing if it correctly computed the values in a known, hand-worked example.  It should also end up a whole lot simpler than you have this test, too.